charms/worker/k8s/src/charm.py

```diff
@@ -1061,14 +1120,102 @@ def _get_external_kubeconfig(self, event: ops.ActionEvent):
+        """Get the most public address either from external load balancer or from juju.
+
+        If the external load balancer is available and the unit is a control-plane unit,
+        the external load balancer address will be used. Otherwise, the juju public address
+        will be used.
+        NOTE: Don't ignore the unit's IP in the extra SANs just because there's a load balancer.
+
+        Returns:
+            str: public ip address of the unit
+            None: if the public address is not available
+        """
+        if self.is_control_plane and self.external_load_balancer.is_available:
+            log.info("Using external load balancer address as the public address")
+            return self._get_external_load_balancer_address()
+        log.info("Using juju public address as the public address")
+        return _get_juju_public_address()
+
+    def _get_external_load_balancer_address(self) -> Optional[str]:
+        """Get the external load balancer address.
+
+        Returns:
+            None: If the external load balancer address is not available.
+            str: The external load balancer address.
+        """
+        if not self.is_control_plane:
+            log.error("External load balancer address is only available for control-plane units")
+            return None
+
+        response = self.external_load_balancer.get_response(EXTERNAL_LOAD_BALANCER_RESPONSE_NAME)
+
+        if not response:
+            log.error("No response from external load balancer when trying to get address")
+            return None
+        if response.error:
+            log.error(
+                "Error from external load balancer when trying to get address: %s",
+                response.error,
+            )
+            return None
+
+        return response.address
+
+    @on_error(
+        ops.WaitingStatus("Ensuring SANs are up-to-date"),
+        InvalidResponseError,
+        K8sdConnectionError,
+    )
+    def _ensure_sans(self):
+        """Ensure the extra SANs are up-to-date.
+
+        This method checks if the extra SANs are present in the API server certificate.
+        If they are not, the certificates are refreshed with the new SANs.
+        """
+        if not self.is_control_plane:
+            return
+
+        extra_sans = self._get_extra_sans()
+        if not extra_sans:
+            log.info("No extra SANs to update")
+            return
+
+        dns_sans, ip_addresses = get_certificate_sans(APISERVER_CERT)
+        ip_addresses = [str(ip) for ip in ip_addresses]
+        all_cert_sans = dns_sans + ip_addresses
+
+        for san in extra_sans:
+            if san not in all_cert_sans:
+                log.info(f"{san} not in cert SANs, refreshing certs with new SANs: {extra_sans}")
+                status.add(ops.MaintenanceStatus("Refreshing Certificates"))
+                self.api_manager.refresh_certs(extra_sans)
+                log.info("Certificates have been refreshed")
```

Comment:
This calls `refresh_certs` in a loop, once for each missing SAN. Therefore, the certificates will be refreshed multiple times unnecessarily:
```suggestion
        missing_sans = [san for san in extra_sans if san not in all_cert_sans]
        if missing_sans:
                log.info("Missing SANs found, refreshing certificates.")
                status.add(ops.MaintenanceStatus("Refreshing Certificates"))
                self.api_manager.refresh_certs(extra_sans)
                log.info("Certificates have been refreshed")
```

Reply:
That's a really good catch. Thanks!